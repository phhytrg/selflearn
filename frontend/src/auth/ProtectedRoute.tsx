import { Link, Navigate, Outlet } from 'react-router-dom';
import { useAuth } from './hooks/useAuth';
import { Button, Drawer } from 'antd';
import { useState } from 'react';
import { MenuOutlined } from '@ant-design/icons';
import { useRoles } from '@/shared/hooks/useAuthority';

export const ProtectedRoute = () => {
  const { isAdmin } = useRoles();
  const [open, setOpen] = useState(false);
  const { user, logout } = useAuth();

  const showDrawer = () => {
    setOpen(true);
  };

  const onClose = () => {
    setOpen(false);
  };

  if (!user) {
    //Try to get user from local storage
    return <Navigate to="/login" />;
  }

  return (
    <div className="flex flex-col">
      <Button
        className="ml-auto"
        aria-label="menu"
        icon={<MenuOutlined />}
        onClick={showDrawer}
      />
      <Drawer
        data-testid="drawer"
        onClose={onClose}
        open={open}
        styles={{
          body: {
            display: 'flex',
            flexDirection: 'column',
            gap: '8px',
          },
        }}
      >
        <Link to={'/Home'}>Home</Link>
        {!isAdmin || <Link to={'/Admin'}>Admin</Link>}
        <a onClick={logout}>Log out</a>
      </Drawer>
      <Outlet />
    </div>
  );
};
