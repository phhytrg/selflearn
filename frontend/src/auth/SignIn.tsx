import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { Button, Input, Modal, Spin } from 'antd';
import { useAuth } from './hooks/useAuth';

const LoginPage = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const navigate = useNavigate();
  const { login, user } = useAuth();
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (user) {
      navigate('/', { replace: true });
    }
  }, [user, navigate]);

  const handleLogin = async (e: any) => {
    e.preventDefault();
    try {
      setIsLoading(true);
      const res = await login({
        username: email,
        password,
      });
      console.log(res);
    } catch (e: any) {
      Modal.error({
        title: 'Error',
        content: (
          <div data-testid="error-modal">
            {JSON.stringify(e.response?.data) || e.message}
          </div>
        ),
      });
      return;
    } finally {
      setIsLoading(false);
    }
    navigate('/', { replace: true });
  };

  return (
    <div className="flex flex-col gap-4">
      <div>
        <label htmlFor="email">Email:</label>
        <Input
          id="email"
          type="text"
          name="email"
          aria-label="email"
          value={email}
          onChange={(e) => setEmail(e.target.value)}
        />
      </div>
      <div>
        <label htmlFor="password">Password:</label>
        <Input
          id="password"
          type="password"
          name="password"
          aria-label="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
        />
      </div>
      <Button
        type="primary"
        onClick={handleLogin}
        name="signIn"
        aria-label="signIn"
        disabled={isLoading}
      >
        {isLoading ? <Spin /> : `Log In`}
      </Button>
      <Button
        type="default"
        name="signUp"
        aria-label="signUp"
        onClick={() => {
          navigate('/signup', { replace: true });
        }}
      >
        Sign Up
      </Button>
    </div>
  );
};

export default LoginPage;
