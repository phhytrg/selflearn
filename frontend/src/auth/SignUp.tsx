
import { Button, Input } from 'antd';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useAuth } from './hooks/useAuth';
import { authApi } from '@/shared/apis';

export const SignUpPage = () => {
  const {user} = useAuth();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [confirmPassword, setConfirmPassword] = useState('');
  const navigate = useNavigate();

  useEffect(() => {
    if (user) {
      navigate('/', { replace: true });
    }
  }, [user, navigate]);

  const handleSignUp = async (e: any) => {
    e.preventDefault();
    authApi.register(email, password, confirmPassword);
    navigate('/login', { replace: true });
  };

  return (
    <div className="flex flex-col gap-4">
      <h1>Sign Up</h1>
      <Input
        type="text"
        placeholder="Email"
        value={email}
        onChange={(e) => {
          setEmail(e.target.value);
        }}
      />
      <Input
        type="password"
        placeholder="Password"
        value={password}
        onChange={(e) => {
          setPassword(e.target.value);
        }}
      />
      <Input
        type="password"
        placeholder="Confirm Password"
        value={confirmPassword}
        onChange={(e) => {
          setConfirmPassword(e.target.value);
        }}
      />
      <Button type="primary" onClick={handleSignUp}>
        Sign Up
      </Button>
    </div>
  );
};
