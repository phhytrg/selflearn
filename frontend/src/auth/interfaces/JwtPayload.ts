export interface JwtPayload {
  sub: string;
  email: string;
  roles: {
    authority: string;
  }[];
  iat: number;
  exp: number;
}
