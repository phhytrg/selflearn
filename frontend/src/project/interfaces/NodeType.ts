export enum NodeType {
  Tree = 'tree',
  Blob = 'blob',
}