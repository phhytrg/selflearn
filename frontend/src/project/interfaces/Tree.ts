export interface TreeNode {
  path: string;
  mode: string;
  type: string;
  sha: string;
  url: string;
  children: TreeNode[];
}

export interface Tree {
  sha: string;
  url: string;
  tree: TreeNode[];
  truncated: boolean;
}
