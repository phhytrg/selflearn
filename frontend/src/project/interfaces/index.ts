export * from './FileContent';
export * from './NodeType';
export * from './Tree';