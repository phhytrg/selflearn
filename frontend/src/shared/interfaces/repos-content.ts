export interface ReposContent {
  content: string
}