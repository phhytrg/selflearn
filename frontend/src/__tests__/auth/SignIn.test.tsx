import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  MockedObject,
  vi,
} from 'vitest';
import {
  cleanup,
  fireEvent,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
  within,
} from '@testing-library/react';
import LoginPage from '@/auth/SignIn';
import {
  AuthContext,
  AuthContextProps,
  AuthProvider,
} from '@/auth/providers/AuthProvider';
import { JwtPayload } from '@/auth/interfaces/JwtPayload';
import {
  BrowserRouter,
  createMemoryRouter,
  RouterProvider,
} from 'react-router-dom';
import userEvent from '@testing-library/user-event';
import { handlers as authHandlers } from '../msw/mocks/auth.handlers';
import { server } from '../msw/worker';
import routesConfig from '@/routes';

// Faker Js for generating fake data
// Mws for mocking the api calls
// Vitest for testing
// React rendering library for rendering the components

// vi.mock('@/auth/hooks/useAuth', async () => {
//   return {
//     useAuth: () => {
//       return {
//         user: null,
//         login: vi.fn(
//           async ({ username }: { username: string; password: string }) => {
//             const user: JwtPayload = {
//               sub: '1',
//               roles: [{ authority: 'ROLE_READER' }],
//               email: username,
//               iat: 0,
//               exp: 0,
//             };
//             return user;
//           },
//         ),
//         logout: vi.fn(() => {
//           return null;
//         }),
//       };
//     },
//   };
// });

server.use(...authHandlers);

vi.mock('antd', async (importOriginal) => {
  const mod = await importOriginal<typeof import('antd')>();
  return {
    ...mod,
    Spin: () => <div aria-label="spin">Loading...</div>,
  };
});

const customRender = (ui: React.ReactNode, { providerProps }) => {
  return render(
    <AuthContext.Provider value={providerProps}>{ui}</AuthContext.Provider>,
    {
      wrapper: BrowserRouter,
    },
  );
};

describe('SignIn Component', () => {
  let providerProps: MockedObject<AuthContextProps>;

  beforeAll(() => {
    // What to do before all tests
    // JSDom does not implement this and an error was being
    // thrown from jest-axe because of it.

    window.getComputedStyle = (elt: Element) => {
      return {
        getPropertyValue: (prop: string) => {
          return '';
        },
      } as CSSStyleDeclaration;
    };
  });

  beforeEach(() => {
    // Thing to do before each tests
    providerProps = {
      user: null,
      login: vi.fn(
        async ({ username }: { username: string; password: string }) => {
          const user: JwtPayload = {
            sub: '1',
            roles: [{ authority: 'ROLE_READER' }],
            email: username,
            iat: 0,
            exp: 0,
          };
          providerProps.user = user;
        },
      ),
      logout: vi.fn(() => {
        providerProps.user = null;
      }),
    };
    customRender(<LoginPage />, { providerProps });
  });

  afterEach(() => {
    // Thing to do after each tests
    cleanup();
  });

  it('should match the snapshot', () => {
    server.use(...authHandlers);
    console.log(server.listHandlers());
    cleanup();
    const { asFragment } = customRender(<LoginPage />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the SignIn component', () => {
    // console.log(server.listHandlers());
    // Sign in component should have email and password fields and a login button, a sign up button
    expect(screen.getByLabelText('Email:')).toBeInTheDocument();
    expect(screen.getByLabelText('Password:')).toBeInTheDocument();
    expect(screen.getByRole('textbox', { name: 'email' })).toBeInTheDocument();
    expect(screen.getByLabelText('password')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'signIn' })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'signUp' })).toBeInTheDocument();
  });

  it('should block sign in button when request is in progress', async () => {
    const buttonElement = screen.getByRole('button', { name: 'signIn' });
    fireEvent.click(buttonElement);
    expect(buttonElement).toBeDisabled();

    await waitForElementToBeRemoved(() => {
      const spinElement = screen.getByLabelText('spin');
      waitFor(() => expect(spinElement).not.toBeInTheDocument());

      return spinElement;
    });
    expect(buttonElement).toBeEnabled();
  });

  const handleLogin = vi.fn();
  it('should call the login function when the form is submitted', async () => {
    const buttonElement = screen.getByRole('button', { name: 'signIn' });
    buttonElement.onclick = handleLogin;
    fireEvent.click(buttonElement);

    expect(handleLogin).toHaveBeenCalled();
  });

  it('should show an error message when the email is invalid', async () => {
    const emailInput = screen.getByRole('textbox', { name: 'email' });
    userEvent.type(emailInput, 'invalid-email');
    const buttonElement = screen.getByRole('button', { name: 'signIn' });
    const passwordElem = screen.getByLabelText('password');
    const emailElem = screen.getByLabelText('email');

    fireEvent.change(emailElem, { target: { value: 'test@gmail.com' } });
    fireEvent.change(passwordElem, { target: { value: 'Test@1234' } });
    fireEvent.click(buttonElement);

    await waitForElementToBeRemoved(() => {
      const spinElement = screen.getByLabelText('spin');
      waitFor(() => expect(spinElement).not.toBeInTheDocument());

      return spinElement;
    });

    // Check if the error message is displayed
  });

  it('should go to sign up page when sign up button is clicked', async () => {
    cleanup();
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ['/home'],
    });
    render(<RouterProvider router={router} />);

    const signUpButton = screen.getByRole('button', { name: 'signUp' });
    fireEvent.click(signUpButton);

    await waitFor(() => {
      expect(
        screen.getByRole('heading', { name: 'Sign Up' }),
      ).toBeInTheDocument();
    });
  });

  it('should show an error message when the password is invalid', async () => {
    cleanup();
    vi.resetModules();
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ['/login'],
    });

    render(
      <AuthProvider>
        <RouterProvider router={router} />
      </AuthProvider>,
    );
    const passwordElem = screen.getByLabelText('password');
    const emailElem = screen.getByLabelText('email');
    const buttonElement = screen.getByRole('button', { name: 'signIn' });

    fireEvent.change(emailElem, { target: { value: 'test@gmail.com' } });
    // Valid password should be Abc@1234
    fireEvent.change(passwordElem, { target: { value: 'abc@1234' } });
    fireEvent.click(buttonElement);

    await waitForElementToBeRemoved(() => {
      const spinElement = screen.getByLabelText('spin');
      waitFor(() => expect(spinElement).not.toBeInTheDocument());

      return spinElement;
    });

    const dialog = await screen.findByRole('dialog');

    expect(dialog).toBeInTheDocument();
    const errorMessage = await within(dialog).findByText(/Error/i);
    expect(errorMessage).toBeInTheDocument();

    // expect(screen.getByTestId('error-modal')).toBeInTheDocument();
  });

  it('should navigate to home page when user is logged in', async () => {
    cleanup();
    const router = createMemoryRouter(routesConfig, {
      initialEntries: ['/home'],
    });

    render(
      <AuthProvider>
        <RouterProvider router={router} />
      </AuthProvider>,
    );
    
    const buttonElement = screen.getByRole('button', { name: 'signIn' });
    const passwordElem = screen.getByLabelText('password');
    const emailElem = screen.getByLabelText('email');

    fireEvent.change(emailElem, { target: { value: 'test@gmail.com' } });
    fireEvent.change(passwordElem, { target: { value: 'Test@1234' } });
    fireEvent.click(buttonElement);

    await waitForElementToBeRemoved(() => {
      const spinElement = screen.getByLabelText('spin');
      waitFor(() => expect(spinElement).not.toBeInTheDocument());

      return spinElement;
    });

    // expect(await screen.findByTestId('error-modal'))?.not.toBeInTheDocument();
    // expect(await screen.findByRole('dialog', {name: /Error/i})).toBeInTheDocument();

    // Check if the user is redirected to the home page
    expect(screen.getByRole('link', { name: 'Home' })).toHaveAttribute(
      'href',
      '/home',
    );
  });

  afterAll(() => {
    console.log('msw handlers: ', server.listHandlers());
    // server.resetHandlers();
  });
});
