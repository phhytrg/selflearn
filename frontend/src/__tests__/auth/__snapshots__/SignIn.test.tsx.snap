// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignIn Component > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4"
  >
    <div>
      <label
        for="email"
      >
        Email:
      </label>
      <input
        aria-label="email"
        class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
        Password:
      </label>
      <input
        aria-label="password"
        class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <button
      aria-label="signIn"
      class="ant-btn css-dev-only-do-not-override-kghr11 ant-btn-primary"
      name="signIn"
      type="button"
    >
      <span>
        Log In
      </span>
    </button>
    <button
      aria-label="signUp"
      class="ant-btn css-dev-only-do-not-override-kghr11 ant-btn-default"
      name="signUp"
      type="button"
    >
      <span>
        Sign Up
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`SignInComponent > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4"
  >
    <div>
      <label
        for="email"
      >
        Email:
      </label>
      <input
        aria-label="email"
        class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
        id="email"
        name="email"
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        for="password"
      >
        Password:
      </label>
      <input
        aria-label="password"
        class="ant-input css-dev-only-do-not-override-kghr11 ant-input-outlined"
        id="password"
        name="password"
        type="password"
        value=""
      />
    </div>
    <button
      aria-label="signIn"
      class="ant-btn css-dev-only-do-not-override-kghr11 ant-btn-primary"
      name="signIn"
      type="button"
    >
      <span>
        Log In
      </span>
    </button>
    <button
      aria-label="signUp"
      class="ant-btn css-dev-only-do-not-override-kghr11 ant-btn-default"
      name="signUp"
      type="button"
    >
      <span>
        Sign Up
      </span>
    </button>
  </div>
</DocumentFragment>
`;
